### PR TITLE
dir: Support finding related refs and dependencies offline

### PR DESCRIPTION
**Pull request by mwleeds**
Here are two commits to fix finding related refs and dependencies in the P2P case.

### CODE REVIEW

**Comment by pwithnall** [comment]
The ‘last discussion for this’ I was thinking about is https://github.com/flatpak/flatpak/issues/1592#issuecomment-382818329.

**Comment by mwleeds** [comment]
> The whole branch could really do with some unit tests.
I agree but I think getting the feature working is more urgent. 3.4 will be here before we know it.

File: common/flatpak-dir.c [hunk 1]
```diff
@@ -235,28 +235,39 @@ flatpak_remote_state_ensure_metadata (FlatpakRemoteState *self,
   return TRUE;
 }
 
-
-char *
+/* Returns TRUE if the ref is found in the summary or cache. out_checksum and
+ * out_variant are not guaranteed to be set even when the ref is found. */
```
[the remark this answers is not in the document]
Comment (mwleeds) [reply common/flatpak-dir.c]: >All the callers of this function which you’ve modified in this commit, and which pass a GError* in to the function call, set the GError in the case that the function returns FALSE but doesn’t set a GError itself.

No, the callers set the error if it hasn't been set and they need it set, regardless of whether TRUE or FALSE was returned.

>I suggest you change this function to always set the GError iff it returns FALSE.

This is already true.

>Guarantee that out_checksum and out_variant are non-NULL iff the function returns TRUE.

I don't understand how this is possible. We don't have the checksum if we found the commit in the xa.cache data.

**Comment by mwleeds** [comment]
> IIRC the outcome of the discussion about this was to look for related ref information in the metadata of the commit containing the app.

Yes, and that should be fixed when #1592 is fixed (hopefully soon).
> If we look for xa.cache in the ostree-metadata ref, peers can’t regenerate that cache (since that would break the signature on ostree-metadata).
> 
> If we look for xa.cache in the summary file, it can be regenerated, but is unsigned.

This is all pretty speculative because we don't have a way for clients to regenerate `xa.cache` without breaking things at the moment. I agree that passing around the whole set of metadata for Flathub could get to be a lot, but the way the code is architected will make it difficult to do anything else at the moment. The remote represented by a `FlatpakRemoteState` object is either a configured remote or a `file://` URI, and usually the former even for P2P installs, so if each peer source were providing an `xa.cache` representing their own refs rather than the ones available on the Internet, there's no single value that would make sense to store in the `metadata` field of the `FlatpakRemoteState` object and use for things like `flatpak_remote_state_lookup_ref`. You'd have to combine them all to get something that makes sense, and it's hard to see how that would work well.
> Why have you gone with xa.cache from ostree-metadata? Am I misremembering the outcome of the last discussion about this? A pointer to that discussion would be helpful (here and in the commit message).

This commit in particular (06fa41e65fa0c21c268ca93fda7805c83c22e916) doesn't use `xa.cache` for finding what related refs are needed. That's something the code already did. This commit only uses `xa.cache` to check if a given related ref is available in a given remote, which is an operation that can't be accomplished by looking at a specific commit's metadata.
> The code in this commit seems to get `xa.cache` from either `summary` or `ostree-metadata`, and that’s not reflected in the commit message.

The commit message doesn't specifically say that the `xa.cache` comes from `ostree-metadata`.

**Comment by mwleeds** [comment]
From IRC:
```
mwleeds: do you understand the reason behind https://github.com/flatpak/flatpak/blob/421b1a953c8578979ecada0da17a8e129a81ca4f/common/flatpak-dir.c#L10952
Because it's weird to look for the ref in the remote but silently ignore the error if it's not there

alexlarsson: related refs are often not in the remote
optional plugins and whatnot
The whole idea of them is "install if they exist in the same remote, do nothing otherwise"
```

**Comment by mwleeds** [comment]
(That IRC discussion validates the current approach on this branch.)

**Comment by pwithnall** [comment]
OK, I’m convinced. Thanks for bearing with me.

@rh-atomic-bot r+ 26dd532d31f0d0f3730a050f844a2920f0e3fa9e

**Pull request closed without merging**